### PR TITLE
[REM3-340] Attempt to write directly rather than resuming writes

### PR DESCRIPTION
**Pull request by stuartwdouglas**

### CODE REVIEW

File: src/main/java/org/jboss/remoting3/remote/RemoteConnection.java [hunk 1]
```diff
@@ -21,6 +21,7 @@
 import java.io.IOException;
 import java.net.InetSocketAddress;
 import java.nio.ByteBuffer;
+import java.rmi.Remote;
```
Comment (dmlloyd) [inline src/main/java/org/jboss/remoting3/remote/RemoteConnection.java]: I guess that this was inadvertent?

File: src/main/java/org/jboss/remoting3/remote/RemoteConnection.java [hunk 2]
```diff
@@ -374,7 +375,11 @@ public void send(final Pooled<ByteBuffer> pooled, final boolean close) {
                         queue.add(pooled);
                         free = false;
                         if (empty) {
-                            connection.getSinkChannel().resumeWrites();
+                            //if there was no data previously queued we add a task to attempt to write the
+                            //data, and resume writes if it fails. This means that if we have multiple messages
+                            //that are to be send they can all be batched into a single write, while also
+                            //preventing a resumeWrites unless it is actually required
+                            connection.getIoThread().execute(flushTask);
```
Comment (dmlloyd) [inline src/main/java/org/jboss/remoting3/remote/RemoteConnection.java]: Does this really make a difference?

Comment (stuartwdouglas) [reply src/main/java/org/jboss/remoting3/remote/RemoteConnection.java]: Yes, resumeReads/resumeWrites looks cheap, but the actual changes get queued, and then get processed as part of the select() call, so if you resume then suspend immediately both ops end up in the queue and even though the net result is zero change NIO still processes them which is not free. 

In terms of the batching that was the original reason why this code is like it is, it is for the server -> server use case where you can potentially have lots of threads using the same connection, and this means that multiple messages can be part of a single write call.

Comment (fl4via) [reply src/main/java/org/jboss/remoting3/remote/RemoteConnection.java]: @dmlloyd this change got us an improvement of 2ms per stateless session bean call on a simple scenario we've been benchmarking.

**Pull request merged**